File: backend/govinda-common/src/main/java/net/voytrex/govinda/common/security/SecurityConfig.java
/*
 * Govinda ERP - Security Configuration
 * Copyright 2026 Voytrex
 * SPDX-License-Identifier: Apache-2.0
 */

package net.voytrex.govinda.common.security;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.web.AuthenticationEntryPoint;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.web.servlet.config.annotation.InterceptorRegistry;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

/**
 * Security configuration for Govinda ERP.
 *
 * Implements JWT-based authentication with:
 * - Stateless session management
 * - Role-based access control (RBAC)
 * - Multi-tenant support
 * - Swagger UI authentication support
 */
@Configuration
@EnableWebSecurity
@EnableMethodSecurity(prePostEnabled = true)
public class SecurityConfig implements WebMvcConfigurer {
    private final boolean swaggerUiEnabled;
    private final JwtAuthenticationFilter jwtAuthenticationFilter;
    private final TenantContextResolver tenantContextResolver;

    public SecurityConfig(
        @Value("${springdoc.swagger-ui.enabled:true}") boolean swaggerUiEnabled,
        JwtAuthenticationFilter jwtAuthenticationFilter,
        TenantContextResolver tenantContextResolver
    ) {
        this.swaggerUiEnabled = swaggerUiEnabled;
        this.jwtAuthenticationFilter = jwtAuthenticationFilter;
        this.tenantContextResolver = tenantContextResolver;
    }

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }

    @Bean
    public AuthenticationEntryPoint authenticationEntryPoint() {
        return (request, response, authException) -> {
            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
            response.setContentType("application/json");
            response.getWriter().write(
                "{\"errorCode\":\"AUTHENTICATION_REQUIRED\",\"message\":\"Authentication required\",\"path\":\""
                    + request.getRequestURI() + "\"}"
            );
        };
    }

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
        http
            .csrf(csrf -> csrf.disable())
            .sessionManagement(session -> session.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
            .authorizeHttpRequests(authz -> {
                if (swaggerUiEnabled) {
                    authz.requestMatchers(
                        "/swagger-ui/**",
                        "/swagger-ui.html",
                        "/v3/api-docs/**",
                        "/api-docs/**",
                        "/swagger-resources/**",
                        "/webjars/**"
                    ).permitAll();
                }
                authz
                    .requestMatchers("/api/v1/auth/login").permitAll()
                    .requestMatchers("/actuator/health").permitAll()
                    .requestMatchers("/api/**").authenticated()
                    .requestMatchers("/").permitAll()
                    .anyRequest().authenticated();
            })
            .exceptionHandling(exceptions -> exceptions
                .authenticationEntryPoint(authenticationEntryPoint())
            )
            .addFilterBefore(jwtAuthenticationFilter, UsernamePasswordAuthenticationFilter.class)
            .formLogin(form -> form.disable())
            .httpBasic(basic -> basic.disable());

        return http.build();
    }

    @Override
    public void addInterceptors(InterceptorRegistry registry) {
        registry.addInterceptor(tenantContextResolver);
    }
}

File: backend/govinda-common/src/main/kotlin/net/voytrex/govinda/common/security/SecurityConfig.kt
/*
 * Govinda ERP - Security Configuration
 * Copyright 2024 Voytrex
 * SPDX-License-Identifier: Apache-2.0
 */

package net.voytrex.govinda.common.security

import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.web.SecurityFilterChain

/**
 * Security configuration for Govinda ERP.
 *
 * Swagger UI and API docs are open by default (no authentication required).
 * This is the recommended approach for development environments.
 *
 * For production, you can:
 * 1. Disable Swagger UI entirely via springdoc.swagger-ui.enabled=false
 * 2. Add authentication/authorization to protect Swagger UI
 * 3. Use a reverse proxy with authentication in front of the application
 */
@Configuration
@EnableWebSecurity
class SecurityConfig(
    @Value("\${springdoc.swagger-ui.enabled:true}")
    private val swaggerUiEnabled: Boolean
) {

    @Bean
    fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
        http
            .csrf { it.disable() } // Disable CSRF for stateless API
            .sessionManagement { it.sessionCreationPolicy(SessionCreationPolicy.STATELESS) }
            .authorizeHttpRequests { authz ->
                authz
                    // Swagger UI and API docs - open for development
                    // Best practice: Open in dev, protected/disabled in production
                    .requestMatchers(
                        "/swagger-ui/**",
                        "/swagger-ui.html",
                        "/v3/api-docs/**",
                        "/api-docs/**",
                        "/swagger-resources/**",
                        "/webjars/**"
                    ).permitAll()
                    // Actuator health endpoint - open for monitoring
                    .requestMatchers("/actuator/health").permitAll()
                    // All other endpoints - currently open (TODO: implement authentication)
                    // API endpoints should require X-Tenant-Id header validation
                    .anyRequest().permitAll()
            }
            // Disable default login form and basic auth
            .formLogin { it.disable() }
            .httpBasic { it.disable() }

        return http.build()
    }
}
